src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl

```diff
@@ -0,0 +1,81 @@
+export TDLearner
+
+using LinearAlgebra: dot
+using Distributions: pdf
+import Base.push!
+
+using ReinforcementLearningCore: AbstractLearner, TabularApproximator
+using Flux
+
+"""
+    TDLearner(;approximator, γ=1.0, method, n=0)
+
+Use temporal-difference method to estimate state value or state-action value.
+
+# Fields
+- `approximator` is `<:TabularApproximator`.
+- `γ=1.0`, discount rate.
+- `method`: only `:SARS` (Q-learning) is supported for the time being.
+- `n=0`: the number of time steps used minus 1.
+"""
+Base.@kwdef mutable struct TDLearner{M,A} <: AbstractLearner where {A<:TabularApproximator,M<:Symbol}
+    approximator::A
+    γ::Float64 = 1.0
+    n::Int = 0
+
+    function TDLearner(approximator::A, method::Symbol; γ=1.0, n=0) where {A<:TabularApproximator}
+        if method ∉ [:SARS]
+            @error "Method $method is not supported"
+        else
+            new{method, A}(approximator, γ, n)
+        end
+    end
+end
+
+RLCore.forward(L::TDLearner, s) = RLCore.forward(L.approximator, s)
+RLCore.forward(L::TDLearner, s, a) = RLCore.forward(L.approximator, s, a)
+
+Q(app::TabularApproximator, s, a) = RLCore.forward(app, s, a)
+Q(app::TabularApproximator, s) = RLCore.forward(app, s)
+
+"""
+    Q!(app::TabularApproximator, s::Int, s_plus_one::Int, a::Int, α::Float64, π_::Float64, γ::Float64)
```

Comment:
I'm not a fan of naming the Bellman update function `Q!`. I'd rather go with a name which is more descriptive of the function. Like `bellman_update!` or `update_Q!`, or whatever.